### ISSUE TITLE
/test: resolve pending TODOs

### ISSUE THREAD

**Issue by tanjunchen** [kubernetes/kubernetes#86756]
run following command in root of kubernetes
```
grep -rs "TODO" --exclude-dir="vendor/" --exclude-dir="third_party/" --exclude-dir="_output/" --exclude-dir="build/"  | wc -l
2930
```
run  following command in /test of kubernetes
```
grep -rs "TODO"  | wc -l
437
```
There are some TODO in /test , I guess that we can git rid of some of them if posssible.
I open an issue to record him, and advocate the removal or consummationof some TODO. 
If you have any comments and ideas, please leave  messages under the issue, thank you.

**How to resovle pending TODO?**
If you are interested in solving a TODO, you may refer to the following steps:
First of all, understand the reason and background of adding this TODO.

Secondly, You can check the owner of TODO through git blame or github logs. If you need to get more information about this TODO, you can ask them through slack, but first, it's best to do your own research.

Thirdly,If  you have solved a TODO, you are welcome to submit a pr. If you list the relevant background,reasons,and have a better solution, then it is best. And leaving the comment ```/assign owner ```. If there is no response for a long time (one week), then you can go to the relevant owner file @other people.

finnaly, If you and the person who left this TODO feel that it cannot be handled temporarily, then comment that the TODO still needs to be retained.

/sig testing
/area e2e-test-framework
/kind cleanup
/priority backlog

- [ ] e2e_node/util.go:// TODO(random-liu): Get this automatically from kubelet flag.
- [ ] e2e_node/util.go:// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
- [ ] e2e_node/environment/conformance.go:// TODO: Should we write an e2e test for this?
- [ ] e2e_node/node_problem_detector_linux.go:			// TODO(random-liu): Now node condition could be arbitrary string, consider whether we need to
- [ ] e2e_node/node_problem_detector_linux.go:			// TODO: remove hardcoded kubelet volume directory path
- [ ] e2e_node/volume_manager_test.go:											// TODO: remove hardcoded kubelet volume directory path
- [ ] e2e_node/garbage_collector_test.go:	//TODO (dashpole): Once dynamic config is possible, test different values for maxPerPodContainer and maxContainers
- [ ] e2e_node/jenkins/e2e-node-jenkins.sh:# TODO converge build steps with hack/build-go some day if possible.
- [ ] e2e_node/quota_lsci_test.go:	// TODO: remove hardcoded kubelet volume directory path
- [ ] e2e_node/quota_lsci_test.go:			// TODO: remove hardcoded kubelet volume directory path
- [ ] e2e_node/system/specs/gke.yaml:    # TODO(yguo0905): Add a description for BCACHE.
- [ ] e2e_node/system/specs/gke.yaml:# TODO(yguo0905): Figure out whether watchdog is required.
- [ ] e2e_node/summary_test.go:					// TODO(#28407): Handle non-eth0 network interface names.
- [ ] e2e_node/summary_test.go:					// TODO(#28393): Test secret volumes
- [ ] e2e_node/summary_test.go:					// TODO(#28394): Test hostpath volumes
- [ ] e2e_node/density_test.go:		// TODO(coufon): the values are generous, set more precise limits with benchmark data
- [ ] e2e_node/density_test.go:	// TODO(coufon): is there any way to not hard code this?
- [ ] e2e_node/density_test.go:	// TODO(coufon): in the test we found kubelet starts while it is busy on something, as a result 'syncLoop'
- [ ] e2e_node/density_test.go:	// TODO(coufon): do not trust 'kubelet' metrics since they are not reset!
- [ ] e2e_node/e2e_node_suite_test.go:// TODO(random-liu): Change the following modes to sub-command.
- [ ] e2e_node/e2e_node_suite_test.go:	// TODO(random-liu): Move kubelet start logic out of the test.
- [ ] e2e_node/e2e_node_suite_test.go:	// TODO(random-liu): Move log fetch logic out of the test.
- [ ] e2e_node/e2e_node_suite_test.go:	// TODO(random-liu): Find who is using flag.Parse() and cause errors and move the following logic
- [ ] e2e_node/e2e_node_suite_test.go:	// TODO(pohly): remove RegisterNodeFlags from test_context.go enable Viper config support here?
- [ ] e2e_node/e2e_node_suite_test.go:			// TODO(random-liu): Consider to chroot the whole test process to make writing
- [ ] e2e_node/e2e_node_suite_test.go:	// TODO(yifan): Temporary workaround to disable coreos from auto restart
- [ ] e2e_node/e2e_node_suite_test.go:// TODO(random-liu): Using dynamic kubelet configuration feature to
- [ ] e2e_node/remote/node_conformance.go:// TODO(random-liu): Get rid of this once we switch to cluster e2e node bootstrap script.
- [ ] e2e_node/remote/node_conformance.go:// TODO(random-liu): Switch to use standard node bootstrap script.
- [ ] e2e_node/remote/node_e2e.go:	// TODO: Register flags.
- [ ] e2e_node/remote/node_e2e.go:	// TODO(random-liu): Do this in cloud init after we remove containervm test.
- [ ] e2e_node/remote/utils.go:	// TODO: consider calling bootstrap script to configure host based on OS
- [ ] e2e_node/remote/remote.go:// TODO(random-liu): junitFilePrefix is not prefix actually, the file name is junit-junitFilePrefix.xml. Change the variable name.
- [ ] e2e_node/remote/remote.go:	// TODO: This is a best-effort, temporary hack that only works for
- [ ] e2e_node/remote/remote.go:// TODO(random-liu): Use the log-dump script in cluster e2e.
- [ ] e2e_node/services/internal_services.go:	// TODO(random-liu): Use a loop to stop all services after introducing
- [ ] e2e_node/services/kubelet.go:// TODO(random-liu): Replace this with standard kubelet launcher.
- [ ] e2e_node/services/kubelet.go:		// TODO(random-liu): Get rid of this docker specific thing.
- [ ] e2e_node/services/logs.go:	// TODO(random-liu): Make docker.log non-required.
- [x] e2e_node/apparmor_test.go:	// TODO(random-liu): The test is run as root now, no need to use sudo here.
- [ ] e2e_node/apparmor_test.go:	// TODO(tallclair): Pass this through the image setup rather than hardcoding.
- [ ] e2e_node/runner/remote/run_remote.go:// TODO(coufon): replace 'image' with 'node' in configurations
- [ ] e2e_node/runner/remote/run_remote.go:	// TODO: Add subcommand for node soaking, node conformance, cri validation.
- [ ] e2e_node/runner/remote/run_remote.go:	// TODO(random-liu): Extract out and unify log collection logic with cluste e2e.
- [ ] e2e_node/runner/remote/run_remote.go:		// TODO(random-liu): Remove the docker version check. Use some other command to check
- [ ] e2e_node/node_container_manager_test.go:	// TODO: Update cgroupManager to expose a Status interface to get current Cgroup Settings.
- [ ] e2e_node/image_list.go:	// TODO(random-liu): Use docker client to get rid of docker binary dependency.
- [ ] e2e_node/container_manager_test.go:				// TODO: Test the oom-score-adj logic for burstable more accurately.
- [ ] e2e_node/gpu_device_plugin_test.go:// TODO make this generic and not linked to COS only
- [ ] e2e_node/device_plugin_test.go:	// sampleDevicePluginDSYAML is the path of the daemonset template of the sample device plugin. // TODO: Parametrize it by making it a feature in TestFramework.
- [ ] e2e_node/device_plugin_test.go:			// TODO(vikasc): Instead of hard-coding number of devices, provide number of devices in the sample-device-plugin using configmap
- [ ] e2e_node/conformance/build/Makefile:# TODO(random-liu): Add this into release progress.
- [ ] e2e_node/conformance/run_test.sh:# TODO(random-liu): Use standard installer to install kubelet.
- [ ] e2e_node/conformance/run_test.sh:# TODO(random-liu): Use standard tool to start kubelet in production way (such
- [ ] integration/deployment/deployment_test.go:// TODO: drop the rollback portions of this test when extensions/v1beta1 is no longer served
- [ ] integration/scale/scale_test.go:	// TODO: Disable logging here until we resolve teardown issues which result in
- [ ] integration/examples/apiserver_test.go:	// TODO figure out how to turn on enough of services and dns to run more
- [ ] integration/apiserver/max_request_body_bytes_test.go:		// TODO(liggitt): expect bad request on deep nesting, rather than success on dropped unknown field data
- [ ] integration/apiserver/apply/apply_test.go:// TODO: make a set of test cases in an easy-to-consume place (separate package?) so it's easy to test in both integration and e2e.
- [ ] integration/apiserver/print_test.go:// TODO (soltysh): this list has to go down to 0!
- [ ] integration/apiserver/print_test.go:	"FlowSchema",                 // TODO(yue9944882): remove this comment by merging print-handler for flow-control API
- [ ] integration/apiserver/print_test.go:	"PriorityLevelConfiguration", // TODO(yue9944882): remove this comment by merging print-handler for flow-control API
- [ ] integration/apiserver/print_test.go:		// we do not care about internal objects or lists // TODO make sure this is always true
- [ ] integration/apiserver/apiserver_test.go:				// TODO: this should be a more specific error
- [ ] integration/apiserver/apiserver_test.go:				// TODO: this should be a more specific error
- [ ] integration/apiserver/apiserver_test.go:				// TODO: this should be a more specific error
- [ ] integration/apiserver/apiserver_test.go:				// TODO: this should be a more specific error
- [ ] integration/framework/util.go:// TODO: This file can potentially be moved to a common place used by both e2e and integration tests.
- [ ] integration/framework/util.go:	// TODO: Create a namespace with a given basename.
- [ ] integration/framework/util.go:			// TODO: Once we start creating namespaces, switch to GenerateName.
- [ ] integration/framework/util.go:	// TODO: Remove all resources from a given namespace once we implement CreateTestingNamespace.
- [ ] integration/framework/master_utils.go:	// TODO: Add configs for endpoints controller, scheduler etc
- [ ] integration/framework/master_utils.go:	// TODO have this start method actually use the normal start sequence for the API server
- [ ] integration/framework/master_utils.go:	// TODO: get rid of these tests or port them to secure serving
- [ ] integration/framework/master_utils.go:	// TODO: get rid of these tests or port them to secure serving
- [ ] integration/framework/etcd.go:	// TODO: Check for valid etcd version.
- [ ] integration/client/client_test.go:	// TODO: Re-enable this test when we get #6059 resolved.
- [ ] integration/scheduler_perf/scheduler_perf_types.go:	// TODO: Other predicates and priorities to be added here.
- [ ] integration/scheduler_perf/scheduler_test.go:		// TODO: this needs to be configurable.
- [ ] integration/scheduler_perf/scheduler_test.go:		// TODO Find out why sometimes there may be scheduling blips in the beginning.
- [ ] integration/scheduler_perf/scheduler_test.go:	} // TODO: other predicates/priorities will be processed in subsequent if statements or a switch:).
- [ ] integration/scheduler_perf/scheduler_test.go://TODO: As of now, this function is not doing anything except for reading input values to priority structs.
- [ ] integration/serviceaccount/service_account_test.go:	// TODO: remove rootClient after we refactor pkg/admission to use the clientset.
- [ ] integration/garbagecollector/garbage_collector_test.go:// TODO: Consider how this could be represented with table-style tests (e.g. a
- [ ] integration/ipamperf/ipam_test.go:		// TODO (#61854) find why flakiness is caused by etcd connectivity before enabling always
- [ ] integration/master/secrets_transformation_test.go:		// TODO: add secretbox
- [ ] integration/master/kube_apiserver_test.go:	// TODO(roycaihw): expose response header in rest client and utilize etag here
- [ ] integration/volumescheduling/volume_binding_test.go:	// TODO: remove when max volume limit allows setting through storageclass
- [ ] integration/volumescheduling/volume_binding_test.go:	workqueue.ParallelizeUntil(context.TODO(), 16, podVolumesCount, func(i int) {
- [ ] integration/volumescheduling/volume_binding_test.go:	workqueue.ParallelizeUntil(context.TODO(), 16, podLimit, func(i int) {
- [ ] integration/volumescheduling/volume_binding_test.go:	workqueue.ParallelizeUntil(context.TODO(), 16, len(pods), func(i int) {
- [ ] integration/volumescheduling/volume_binding_test.go:	workqueue.ParallelizeUntil(context.TODO(), 16, len(pvcs), func(i int) {
- [ ] integration/volumescheduling/volume_binding_test.go:		workqueue.ParallelizeUntil(context.TODO(), 16, len(pvs), func(i int) {
- [ ] integration/apimachinery/watch_restart_test.go:			// TODO: get this lower once we figure out how to extend ETCD cache
- [ ] integration/scheduler/framework_test.go:	// TODO: register more plugin which would trigger un-reserve plugin
- [ ] integration/scheduler/framework_test.go:// TODO: refactor it to a more generic functions that accepts all kinds of Plugins as arguments
- [ ] integration/scheduler/taint_test.go:			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
- [ ] integration/auth/accessreview_test.go:// TODO(etune): remove this test once a more comprehensive built-in authorizer is implemented.
- [ ] integration/auth/node_test.go:	//TODO(mikedanese): integration test node restriction of TokenRequest
- [ ] integration/auth/auth_test.go:		// TODO: test .../watch/..., which doesn't end before the test timeout.
- [ ] integration/auth/auth_test.go:		// TODO: figure out how to create a node so that it can successfully proxy/redirect.
- [ ] integration/auth/auth_test.go:// TODO(etune): write a fuzz test of the REST API.
- [ ] integration/auth/auth_test.go:// TODO(etune): remove this test once a more comprehensive built-in authorizer is implemented.
- [ ] cmd/apps.sh:  # TODO: test robust scaling in an e2e.
- [x] cmd/apps.sh:  #TODO(mortent): Remove this workaround when ReplicaSet bug described in issue #69376 is fixed
- [ ] cmd/legacy-script.sh:# TODO: Remove this function when we do the retry inside the kubectl commands. See #15333.
- [ ] cmd/legacy-script.sh:  # TODO: we need to note down the current default namespace and set back to this
- [ ] typecheck/main.go:	// TODO(rmmh): Customize our forked srcimporter to do this better.
- [ ] images/README.md:TODO: Once [Centralization part 4](https://github.com/kubernetes/kubernetes/pull/81226) merges, the paragraph
- [ ] images/README.md:TODO: Once [Centralization part 4](https://github.com/kubernetes/kubernetes/pull/81226) merges, the paragraph
- [ ] images/apparmor-loader/loader.go:// TODO: This is copied from k8s.io/kubernetes/pkg/security/apparmor.getLoadedProfiles.
- [ ] images/pets/peer-finder/peer-finder.go:	// TODO: Exit if there's no on-change?
- [ ] images/pets/redis-installer/Dockerfile:# TODO: get rid of bash dependency and switch to plain busybox.
- [ ] images/pets/redis-installer/Dockerfile:# TODO: just use standard redis when there is one for 3.2.0.
- [ ] images/pets/redis-installer/on-start.sh:        # TODO: More restrictive regex?
- [ ] images/pets/zookeeper-installer/install.sh:# TODO: Should dynamic config be tied to the version?
- [ ] images/pets/zookeeper-installer/install.sh:# TODO: This is a hack, netcat is convenient to have in the zookeeper container
- [ ] images/pets/zookeeper-installer/Dockerfile:# TODO: get rid of bash dependency and switch to plain busybox.
- [ ] images/pets/zookeeper-installer/on-start.sh:    # TODO: zkServer-initialize is the safe way to handle changes to datadir
- [ ] images/pets/zookeeper-installer/on-start.sh:# TODO: zkServer-initialize is the safe way to handle changes to datadir
- [ ] images/pets/zookeeper-installer/on-start.sh:# TODO: We shouldn't need to specify the address of the master as long as
- [ ] images/agnhost/crd-conversion-webhook/config.go:		// TODO: uses mutual tls after we agree on what cert the apiserver should use.
- [ ] images/agnhost/nettest/nettest.go:	CmdNettest.Flags().StringVar(&namespace, "namespace", "default", "Namespace of this pod. TODO: kubernetes should make this discoverable.")
- [ ] images/agnhost/webhook/config.go:		// TODO: uses mutual tls after we agree on what cert the apiserver should use.
- [ ] images/agnhost/webhook/main.go:	// TODO: try this library to see if it generates correct json patch
- [ ] e2e/scheduling/ubernetes_lite_volumes.go:		// TODO: SkipUnlessDefaultScheduler() // Non-default schedulers might not spread
- [x] e2e/scheduling/preemption.go:		// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [x] e2e/scheduling/priorities.go:		// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [ ] e2e/scheduling/ubernetes_lite.go:		// TODO: SkipUnlessDefaultScheduler() // Non-default schedulers might not spread
- [x] e2e/scheduling/predicates.go:		// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [ ] e2e/scheduling/predicates.go:// TODO: upgrade calls in PodAffinity tests when we're able to run them
- [ ] e2e/e2e.go:		// TODO: we should probably only be trying to create this directory once
- [ ] e2e/e2e.go:// TODO: dual-stack may support multiple families per service
- [ ] e2e/e2e.go:	// TODO: In large clusters, we often observe a non-starting pods due to
- [ ] e2e/e2e.go:	// TODO: dual-stack
- [ ] e2e/storage/persistent_volumes-local.go:	// TODO: This may not be available/writable on all images.
- [ ] e2e/storage/persistent_volumes-local.go:		// TODO:
- [ ] e2e/storage/persistent_volumes-local.go:	// TODO: hostname may not be the best option
- [ ] e2e/storage/testsuites/provisioning.go:		// TODO: make class creation optional and remove the IsAlreadyExists exception
- [ ] e2e/storage/testsuites/topology.go:	// TODO: scale?
- [ ] e2e/storage/testsuites/base.go:		// TODO: use a deeper directory hierarchy once gubernator
- [ ] e2e/storage/testsuites/base.go:		// TODO(dyzz): Add a dimension to OperationGenerator metrics for
- [ ] e2e/storage/testsuites/volumemode.go:			// TODO: check if it's a CSI volume first
- [ ] e2e/storage/testsuites/subpath.go:			// TODO: This skip should be removed once #61446 is fixed
- [ ] e2e/storage/testsuites/subpath.go:	// TODO: add a test case for the same disk with two partitions
- [ ] e2e/storage/utils/create.go:		// TODO (?): use same logic as framework.go for determining
- [ ] e2e/storage/utils/create.go:		// TODO: avoid hard-coding this special name. Perhaps add a Framework.PredefinedRoles
- [ ] e2e/storage/subpath.go:			// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default container runtime on Windows.
- [ ] e2e/storage/subpath.go:			// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default container runtime on Windows.
- [ ] e2e/storage/subpath.go:			// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default container runtime on Windows.
- [ ] e2e/storage/subpath.go:			// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default container runtime on Windows.
- [ ] e2e/storage/subpath.go:			// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default container runtime on Windows.
- [ ] e2e/storage/flexvolume.go:	// TODO: change this and config-test.sh when default flex volume install path is changed for GCI
- [ ] e2e/storage/persistent_volumes.go://   next pod. Adding concurrency is a TODO item.
- [ ] e2e/storage/persistent_volumes.go:		// TODO: currently a serialized test of each PV
- [ ] e2e/storage/drivers/in_tree.go:	// TODO(mkimuram): cluster-admin gives too much right but system:persistent-volume-provisioner
- [ ] e2e/storage/drivers/in_tree.go:				// TODO: fix iSCSI driver can work with ext3
- [ ] e2e/storage/drivers/in_tree.go:				// TODO: fix rbd driver can work with ext3
- [ ] e2e/storage/drivers/in_tree.go:	// TODO: check if it is correct
- [ ] e2e/storage/drivers/in_tree.go:	// TODO: check if it is correct
- [ ] e2e/storage/external/external.go:		// TODO (?): load from file
- [ ] e2e/storage/external/external.go:	// TODO: strict checking of the file content once https://github.com/kubernetes/kubernetes/pull/71589
- [ ] e2e/storage/external/external.go:	// TODO (?): add support for more volume types
- [ ] e2e/storage/volume_provisioning.go:// TODO: remove when storage.k8s.io/v1beta1 is removed.
- [ ] e2e/apps/rc.go:	// TODO: Maybe switch PodsCreated to just check owner references.
- [ ] e2e/apps/network_partition.go:		// TODO(foxish): Re-enable testing on gce after kubernetes#56787 is fixed.
- [ ] e2e/apps/network_partition.go:			// TODO(foxish): Re-enable testing on gce after kubernetes#56787 is fixed.
- [ ] e2e/apps/job.go:		// TODO revert this back to above when https://github.com/kubernetes/kubernetes/issues/64787 gets fixed
- [ ] e2e/apps/daemon_restart.go:		// TODO: Can it really?
- [x] e2e/apps/daemon_restart.go:		// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [ ] e2e/apps/replica_set.go:	// TODO: Maybe switch PodsCreated to just check owner references.
- [ ] e2e/apps/deployment.go:	// TODO: add tests that cover deployment.Spec.MinReadySeconds once we solved clock-skew issues
- [ ] e2e/apps/statefulset.go:	// TODO: Find a readiness probe for mysql that guarantees writes will
- [ ] e2e/apps/statefulset.go:	// TODO: we won't use "-index" as the name strategy forever,
- [ ] e2e/windows/density.go:		// TODO(coufon): the values are generous, set more precise limits with benchmark data
- [ ] e2e/windows/gmsa_full.go:	// TODO(wk8): we should pin versions.
- [ ] e2e/windows/dns.go:		// TODO: Add more test cases for other DNSPolicies.
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:			// TODO: possibly remove broken node from newNodesSet to prevent removeLabel from crashing.
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:		// TODO: Remove the extra time when GKE restart is fixed.
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:		// TODO: Remove the extra time when GKE restart is fixed.
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:		// TODO(bskiba): Use locations API for all clusters once it's graduated to v1.
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:	// TODO(bskiba): Use an appropriate indicator that the cluster is regional.
- [x] e2e/autoscaling/cluster_size_autoscaling.go:	// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [ ] e2e/autoscaling/cluster_size_autoscaling.go:// TODO: Remove this when the PDB is added for some of those components
- [ ] e2e/autoscaling/horizontal_pod_autoscaling.go:// TODO The use of 3 states is arbitrary, we could eventually make this test handle "n" states once this test stabilizes.
- [ ] e2e/autoscaling/dns_autoscaling.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #55779 is fixed.
- [ ] e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go:				// TODO(bskiba): change default to real namespace name once it is available
- [ ] e2e/node/kubelet_perf.go:	// TODO: Use a more realistic workload
- [ ] e2e/node/kubelet_perf.go:	// TODO(random-liu): Remove the original log when we migrate to new perfdash
- [ ] e2e/node/kubelet_perf.go:		// TODO(#36621): Deprecate this test once we have a node e2e soak
- [ ] e2e/node/crictl.go:			// TODO: Verify the output.
- [ ] e2e/node/security_context.go:		// TODO: port to SecurityContext as soon as seccomp is out of alpha
- [ ] e2e/node/security_context.go:		// TODO: port to SecurityContext as soon as seccomp is out of alpha
- [ ] e2e/node/security_context.go:		// TODO: port to SecurityContext as soon as seccomp is out of alpha
- [ ] e2e/node/security_context.go:		// TODO: port to SecurityContext as soon as seccomp is out of alpha
- [ ] e2e/instrumentation/monitoring/custom_metrics_stackdriver.go:	// TODO: add some events to wait for instead of fixed amount of time
- [ ] e2e/instrumentation/monitoring/custom_metrics_stackdriver.go:// TODO(kawych): migrate this test to new resource model
- [ ] e2e/instrumentation/monitoring/custom_metrics_stackdriver.go:	// TODO: add some events to wait for instead of fixed amount of time
- [ ] e2e/instrumentation/monitoring/accelerator.go:			// TODO: check only for metrics from this cluster
- [ ] e2e/instrumentation/monitoring/custom_metrics_deployments.go:// TODO: Cleanup time series as well
- [ ] e2e/instrumentation/monitoring/stackdriver.go:			// TODO: check only for metrics from this cluster
- [ ] e2e/instrumentation/logging/generic_soak.go:	// scenarios.  TODO jayunit100 add this to the kube CI in a follow on infra patch.
- [ ] e2e/instrumentation/logging/stackdriver/soak.go:			// TODO(instrumentation): Increase to 21 hrs
- [ ] e2e/instrumentation/logging/stackdriver/utils.go:		// TODO(instrumentation): Filter logs from the current project only.
- [ ] e2e/instrumentation/logging/elasticsearch/kibana.go:		// TODO: For now assume we are only testing cluster logging with Elasticsearch
- [ ] e2e/instrumentation/logging/elasticsearch/basic.go:		// TODO: For now assume we are only testing cluster logging with Elasticsearch
- [ ] e2e/testing-manifests/serviceloadbalancer/netexecrc.yaml:          # TODO: use the downward api and get podname instead.
- [ ] e2e/testing-manifests/statefulset/mysql-galera/statefulset.yaml:          # TODO: If docker exec is buggy just use k8s.gcr.io/mysql-healthz:1.0
- [ ] e2e/testing-manifests/storage-csi/external-snapshotter/rbac.yaml:  namespace: default # TODO: replace with the namespace you want for your sidecar
- [ ] e2e/testing-manifests/storage-csi/external-snapshotter/rbac.yaml:  namespace: default # TODO: replace with the namespace you want for your sidecar
- [ ] e2e/testing-manifests/storage-csi/external-snapshotter/rbac.yaml:    namespace: default # TODO: replace with the namespace you want for your sidecar
- [ ] e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-provisioner.yaml:          # TODO: replace with official 1.5.0 release when ready
- [ ] e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-snapshotter.yaml:        # TODO: replace with official 2.0.0 release when ready
- [ ] e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml:          # TODO: replace with official 1.4.0 release when ready
- [ ] e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml:          # TODO: replace with official 2.0.0 release when ready
- [ ] e2e/testing-manifests/flexvolume/dummy-attachable:	# TODO issue 44737 detach is passed PV name, not mount device
- [ ] e2e/testing-manifests/flexvolume/attachable-with-long-mount:        # TODO issue 44737 detach is passed PV name, not mount device
- [ ] e2e/common/util.go:// TODO(random-liu): Change the image puller pod to use similar mechanism.
- [ ] e2e/common/runtime.go:				// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default
- [ ] e2e/common/runtime.go:				// TODO(claudiub): Remove [LinuxOnly] tag once Containerd becomes the default
- [ ] e2e/common/downwardapi_volume.go:// TODO: add test-webserver example as pointed out in https://github.com/kubernetes/kubernetes/pull/5093#discussion-diff-37606771
- [ ] e2e/common/privileged.go:// TODO: Merge with tests in security_context.go
- [ ] e2e/common/kubelet_etc_hosts.go:	// TODO: workaround for https://github.com/kubernetes/kubernetes/issues/34256
- [ ] e2e/common/init_container.go:		// TODO: we may in the future enable resetting initialized = false if the kubelet needs to restart it
- [ ] e2e/common/init_container.go:					// TODO: more conditions
- [ ] e2e/common/host_path.go://TODO : Consolidate this code with the code for emptyDir.
- [ ] e2e/common/host_path.go:		// TODO permission denied cleanup failures
- [ ] e2e/common/host_path.go://TODO: To merge this with the emptyDir tests, we can make source a lambda.
- [ ] e2e/common/container_probe.go:		// TODO: enable this test once the default exec handler supports timeout.
- [ ] e2e/common/pods.go:		// TODO(filbranden): We would like to use a unique service name such as:
- [ ] e2e/common/pods.go:	// TODO(freehan): label the test to be [NodeConformance] after tests are proven to be stable.
- [ ] e2e/ui/dashboard.go:		// TODO(kubernetes/kubernetes#61559): Enable dashboard here rather than skip the test.
- [ ] e2e/cloud/gcp/addon_update.go:// TODO: it would probably be slightly better to build up the objects
- [ ] e2e/cloud/gcp/addon_update.go:// TODO use the ssh.SSH code, either adding an SCP to it or copying files
- [ ] e2e/cloud/gcp/reboot.go:		// TODO(cjcullen) reduce this sleep (#19314)
- [ ] e2e/cloud/gcp/resize_nodes.go:			// TODO(cjcullen) reduce this sleep (#19314)
- [ ] e2e/cloud/gcp/resize_nodes.go:		// TODO: Bug here - testName is not correct
- [ ] e2e/cloud/gcp/node_lease.go:			// TODO(cjcullen) reduce this sleep (#19314)
- [x] e2e/cloud/nodes.go:		// TODO: write a wrapper for ExpectNoErrorWithOffset()
- [ ] e2e/perftype/perftype.go:// TODO(random-liu): Replace this with prometheus' data model.
- [ ] e2e/kubectl/kubectl.go:			// TODO: Generalize integrating debug info into these tests so we always get debug info when we need it
- [ ] e2e/kubectl/kubectl.go:		// TODO: test proxy options (static, prefix, etc)
- [ ] e2e/kubectl/portforward.go:// TODO support other ports besides 80
- [ ] e2e/e2e_test.go:	// TODO: Deprecating repo-root over time... instead just use gobindata_util.go , see #23987.
- [ ] e2e/upgrades/mysql.go:	// TODO: Not sure what the ratio defining a successful test run should be. At time of writing the
- [ ] e2e/upgrades/apps/deployments.go:// TODO: Test that the deployment stays available during master (and maybe
- [ ] e2e/upgrades/apps/replicasets.go:// TODO: Test that the replicaset stays available during master (and maybe
- [ ] e2e/upgrades/cassandra.go:	// TODO(maisem): tweak this value once we have a few test runs.
- [ ] e2e/upgrades/etcd.go:	// TODO(maisem): tweak this value once we have a few test runs.
- [ ] e2e/network/networking_perf.go:	// TODO replace the 1 with the scale option implementation
- [ ] e2e/network/networking_perf.go:	// TODO: Make this a function parameter, once we distribute iperf endpoints, possibly via session affinity.
- [ ] e2e/network/networking_perf.go:// TODO jayunit100 : Retag this test according to semantics from #22401
- [ ] e2e/network/networking_perf.go:// TODO jayunit100 : Retag this test according to semantics from #22401
- [ ] e2e/network/network_tiers.go:// TODO: add retries if this turns out to be flaky.
- [ ] e2e/network/network_tiers.go:// TODO(#51665): remove this helper function once Network Tiers becomes beta.
- [ ] e2e/network/scale/ingress.go:// TODO: Need a better way/format for data output.
- [ ] e2e/network/fixture.go:		// TODO(mikedanese): Wait.
- [ ] e2e/network/ingress.go:	// TODO: write similar tests for nginx, haproxy and AWS Ingress.
- [ ] e2e/network/ingress.go:			// TODO(59778): Check GCE resources specific to this ingress instead of listing all resources.
- [ ] e2e/network/ingress.go:			// TODO(nikhiljindal): Check the instance group annotation value and verify with a multizone cluster.
- [ ] e2e/network/ingress.go:		// TODO: Implement a multizone e2e that verifies traffic reaches each
- [ ] e2e/network/ingress.go:			// TODO(https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/19):
- [ ] e2e/network/ingress.go:			// TODO: This test may fail on other platforms. We can simply skip it
- [ ] e2e/network/proxy.go:				// TODO: below entries don't work, but I believe we should make them work.
- [ ] e2e/network/proxy.go:	// TODO: investigate why it doesn't work on master Node.
- [ ] e2e/network/proxy.go:	// TODO: Change it to test whether all requests succeeded when requests
- [ ] e2e/network/service.go:	// TODO: This timeout should be O(10s), observed values are O(1m), 5m is very
- [ ] e2e/network/service.go:			// TODO: Use exec-over-http via the netexec pod instead of kubectl exec.
- [ ] e2e/network/service.go:			// TODO: simply checking that the retrieved endpoints is a superset
- [ ] e2e/network/service.go:	// TODO: Make it work for all providers.
- [ ] e2e/network/service.go:	// TODO: We get coverage of TCP/UDP and multi-port services through the DNS test. We should have a simpler test for multi-port TCP here.
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:		// TODO: use the ServiceTestJig here
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:	// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:// TODO: Get rid of [DisabledForLargeClusters] tag when issue #56138 is fixed.
- [ ] e2e/network/service.go:		// TODO: We need to attempt to create another service with the previously
- [ ] e2e/network/dns_common.go:				// TODO: Consider scraping logs instead of running a webserver.
- [ ] e2e/network/dns_common.go:	// TODO: probe from the host, too.
- [ ] e2e/network/dns_scale_records.go:		workqueue.ParallelizeUntil(context.TODO(), parallelCreateServiceWorkers, len(services), createService)
- [ ] e2e/network/networking.go:			// TODO: test proxy links here
- [ ] e2e/network/networking.go:		// TODO: this is overkill we just need the host networking pod
- [ ] e2e/network/dual_stack.go:		// TODO (aramase) can switch to new function to get all nodes
- [ ] e2e/network/dual_stack.go:		// TODO (aramase) can switch to new function to get all nodes
- [ ] e2e/network/dual_stack.go:	// TODO (aramase) remove phase 2 tag once phase 2 of dual stack is merged
- [ ] e2e/network/dual_stack.go:// TODO (aramase) move this to https://github.com/kubernetes/utils/blob/master/net/net.go
- [ ] e2e/network/dual_stack.go:// TODO (aramase) move this to https://github.com/kubernetes/utils/blob/master/net/net.go
- [ ] e2e/network/dns.go:		// TODO: Spin up a separate test service and test that dns works for that service.
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Create more endpoints and ensure that multiple A records are returned
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: Validate both IPv4 and IPv6 families for dual-stack
- [ ] e2e/network/dns.go:		// TODO: For dual stack we can run from here two createTargetedProbeCommand()
- [ ] e2e/network/dns.go:		// TODO: Figure out a better way other than checking the actual resolv,conf file.
- [ ] e2e/network/dns.go:		// TODO: Add more test cases for other DNSPolicies.
- [ ] e2e/apimachinery/table_conversion.go:		workqueue.ParallelizeUntil(context.TODO(), 5, 20, func(i int) {
- [ ] e2e/apimachinery/garbage_collector.go:		// TODO: 30s is enough assuming immediate processing of dependents following
- [ ] e2e/apimachinery/garbage_collector.go:	// TODO: this should be an integration test
- [ ] e2e/apimachinery/garbage_collector.go:		// TODO: shorten the timeout when we make GC's periodic API rediscovery more efficient.
- [ ] e2e/apimachinery/garbage_collector.go:	// TODO: should be an integration test
- [ ] e2e/apimachinery/garbage_collector.go:		// TODO: shorten the timeout when we make GC's periodic API rediscovery more efficient.
- [ ] e2e/apimachinery/chunking.go:		workqueue.ParallelizeUntil(context.TODO(), 20, numberOfTotalResources, func(i int) {
- [ ] e2e/apimachinery/crd_conversion_webhook.go:	// TODO: we have to wait for the storage version to become effective. Storage version changes are not instant.
- [ ] e2e/auth/audit.go:// TODO: Get rid of [DisabledForLargeClusters] when feature request #53455 is ready.
- [ ] e2e/auth/audit_dynamic.go:			// TODO(@pbarker): dedupe this with the main audit test once policy functionality is available
- [ ] utils/create_resources.go:// TODO: Refactor common part of functions in this file for generic object kinds.
- [ ] utils/update_resources.go:	// TODO: Try to move this to exponential backoff by modifying scale.Scale().
- [ ] utils/delete_resources.go:// TODO: Refactor common part of functions in this file for generic object kinds.
- [ ] utils/runners.go:		workqueue.ParallelizeUntil(context.TODO(), podCount, podCount, createPodFunc)
- [ ] utils/runners.go:		workqueue.ParallelizeUntil(context.TODO(), 30, podCount, createPodFunc)
- [ ] utils/runners.go:		workqueue.ParallelizeUntil(context.TODO(), count, count, createPodFunc)
- [ ] utils/runners.go:		workqueue.ParallelizeUntil(context.TODO(), 30, count, createPodFunc)
- [ ] utils/runners.go:// TODO: attach secrets using different possibilities: env vars, image pull secrets.
- [ ] utils/runners.go:// TODO: attach configmaps using different possibilities: env vars.
- [ ] e2e_kubeadm/cluster_info_test.go:		//TODO: What else? server?
- [ ] kubemark/start-kubemark.sh:  # TODO(https://github.com/kubernetes/kubernetes/issues/79883): Migrate all components to separate credentials.
- [ ] kubemark/gce/util.sh:    # TODO(mborsz): Figure out some better way to filter out such env variables than
- [ ] kubemark/iks/util.sh:		# TODO: dynamically adjust the number and spec
- [ ] kubemark/resources/manifests/addons/kubemark-rbac-bindings/kubelet-binding.yaml:# TODO: give each kubelet a credential in the system:nodes group with username system:node:<nodeName>,
- [ ] kubemark/resources/manifests/addons/kubemark-rbac-bindings/cluster-autoscaler-binding.yaml:# TODO: Use proper Cluster Autoscaler role (github.com/kubernetes/autoscaler/issues/383)
- [ ] kubemark/resources/start-kubemark-master.sh:		# TODO: Set larger quota to see if that helps with
- [ ] kubemark/resources/start-kubemark-master.sh:	# TODO(shyamjvs): Incorporate network plugin options, etc later.
- [ ] kubemark/resources/start-kubemark-master.sh:# TODO: Get rid of this function when #53321 is fixed.
- [ ] kubemark/resources/start-kubemark-master.sh:    # TODO(#46983): Change this to the ingress controller service account.
- [ ] kubemark/resources/start-kubemark-master.sh:		# TODO: Currently supporting just log backend. Support webhook if needed.

**Comment by neolit123** [comment]
@tanjunchen 
thank you for enumerating these TODOs.

all TODOs in the framework are the responsibility of the "testing commnons" sub project of SIG Testing. "testing commons" created an initiative to c;leanup are refactor the framework to be more easy to consume.

but /test/kubemark or /test/e2e/sig-name are responsibility of other kubernetes SIGs.

my point is that different TODOs have different owners and every single one of them requires investigation.

> Can we delete some of them ?

i don't see anything outstanding that can be just deleted. if want to help "testing commons", you can work on TODOs in the framework, else you can coordinate with the owning SIGs. all of them have slack channels.

cc @andrewsykim @oomichi

**Comment by tanjunchen** [comment]
@neolit123 
Sure , Thanks  for your comments. I Just record these TODOS in /test , my purpose is not to simply delete these TODOs, but we do some research for each TODO to see if they still need to be retained. I estimate that some have been retained for a long time and we need to do some research and investigation on them. I want to help "testing commons" , I will coordinate with them. I will do some work on them , Just like ``` https://github.com/kubernetes/kubernetes/issues/86721``` . 
Can we separate them? Such as package name or group, they are a bit more.
Do you have any good opinions about the title? Thanks again.

**Comment by neolit123** [comment]
creating a separate issue for the framework seems good to me:

> /test/e2e/framework: resolve pending TODOs

for the rest it is difficult because these are owned by different SIGs and it might contain logic that only a few people from a SIG understand...
